File: app-components/src/main/java/org/synsystems/onlypass/components/cryptography/credentialhardening/CredentialHardener.java
package org.synsystems.onlypass.components.cryptography.credentialhardening;

import android.support.annotation.NonNull;

import org.synsystems.onlypass.components.cryptography.credentials.InsecureCredential;
import org.synsystems.onlypass.components.cryptography.credentials.SecureCredential;

import io.reactivex.Single;

/**
 * Hardens {@link InsecureCredential}s to produce {@link SecureCredential}s.
 *
 * @param <I>
 *     the type of insecure credentials to be hardened
 * @param <S>
 *     the type of secure credentials produced by hardening
 * @param <P>
 *     the type of parameters for configuration on a per-harden basis
 */
public interface CredentialHardener<
    I extends InsecureCredential,
    S extends SecureCredential,
    P extends HardeningParameters> {

  /**
   * Hardens an insecure credential thereby producing a secure credential. Each call with the same arguments must
   * produce the same result.
   *
   * @param insecureCredential
   *     the credential to harden
   * @param hardeningParameters
   *     parameters that configure the hardening process
   *
   * @return a new single that emits the secure credential
   */
  @NonNull
  public Single<S> hardenCredential(@NonNull I insecureCredential, @NonNull P hardeningParameters);
}

File: app-components/src/main/java/org/synsystems/onlypass/components/cryptography/credentialhardening/CredentialHardener.kt
package org.synsystems.onlypass.components.cryptography.credentialhardening

import io.reactivex.Single
import org.synsystems.onlypass.components.cryptography.credentials.InsecureCredential
import org.synsystems.onlypass.components.cryptography.credentials.SecureCredential

/**
 * Hardens [InsecureCredential]s to produce [SecureCredential]s.
 *
 * @param I the type of insecure credentials to be hardened
 * @param S the type of secure credentials produced by hardening
 * @param P the type of parameters used to control hardening
 */
interface CredentialHardener<I : InsecureCredential, S : SecureCredential, P : HardeningParameters> {

  /**
   * Hardens [insecureCredential]. Calls with the same arguments must be idempotent.
   *
   * @param hardeningParameters parameters that configure the hardening process
   *
   * @return the secure credential produced by hardening
   */
  fun hardenCredential(insecureCredential: I, hardeningParameters: P): Single<S>
}

/**
 * Parameters for configuring credential hardening on a per-harden basis.
 */
interface HardeningParameters